Content.Server/_Impstation/Radio/SelfHeadsetSystem.cs

```diff
@@ -0,0 +1,81 @@
+using Content.Server.Emp;
+using Content.Server._Impstation.Radio.Components;
+using Content.Server.Radio.Components;
+using Content.Shared.Radio;
+using Content.Shared.Radio.Components;
+
+namespace Content.Server._Impstation.Radio;
+
+public sealed class SelfHeadsetSystem : EntitySystem
+{
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<SelfHeadsetComponent, EncryptionChannelsChangedEvent>(OnKeysChanged);
+        SubscribeLocalEvent<SelfHeadsetComponent, EmpPulseEvent>(OnEmpPulse);
+        SubscribeLocalEvent<SelfHeadsetComponent, ComponentStartup>(OnAdd);
+
+    }
+
+    /// <summary>
+    /// Used to give an entity access to radio channels when an Encryption Key is inserted, without the need for a headset.
+    /// </summary>
+
+    private void OnKeysChanged(EntityUid uid, SelfHeadsetComponent component, ref EncryptionChannelsChangedEvent args)
+    {
+        UpdateRadioChannels(uid, component, args.Component);
+    }
+
+    private void UpdateRadioChannels(EntityUid uid, SelfHeadsetComponent component, EncryptionKeyHolderComponent? keyHolder = null)
+    {
+        {
+            if (!Resolve(uid, ref keyHolder))
+                return;
+
+            if (keyHolder.Channels.Count == 0)
+                RemCompDeferred<ActiveRadioComponent>(uid);
+            else
+                EnsureComp<ActiveRadioComponent>(uid).Channels = new(keyHolder.Channels);
+        }
+
+        {
+            if (!Resolve(uid, ref keyHolder))
+                return;
+
+            if (keyHolder.Channels.Count == 0)
+                RemCompDeferred<IntrinsicRadioReceiverComponent>(uid);
```

Comment:
just put these with the other channel count check? no need to redo this three times over